tokenscan/scanner.go

```diff
@@ -0,0 +1,221 @@
+package tokenscan
+
+import (
+	"context"
+	"math"
+	"math/rand"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/gocql/gocql"
+
+	"github.com/scylladb/gocqlx"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/scylladb/gocqlx/qb"
+)
+
+const (
+	// parallelism = nodesInCluster * coresInNode * 3
+	defaultParallelism = 72
+)
+
+type tokenRange struct {
+	startRange int64
+	endRange   int64
+}
+
+// SessionBuilder creates a gocql.Session
+type SessionBuilder interface {
+	CreateSession() (*gocql.Session, error)
+}
+
+// ApplyFunc is a callback function that is called with each retrieved scan item.
+// The scan is aborted if the ApplyFunc returns false
+type ApplyFunc func(context.Context, map[string]interface{}) bool
+
+// Scanner is a scylla token range scanner
+type Scanner struct {
+	sessionBuilder SessionBuilder
+	tokenRanges    []*tokenRange
```

Comment:
There is no need to allocate memory for it, check the test producer consumer impl https://github.com/scylladb/gocqlx/blob/a72115182914eced6612440463a3ab40c727c5b4/example_test.go#L549